### ISSUE TITLE
Update ReadMe.md with branching strategy

### ISSUE THREAD

**Issue by rexor12** [rexor12/holobot#16]
Since the branching strategy this repository uses may not be the most straightforward one among those I have seen in other repositories so far, it is definitely worth describing how it is done, in a well-formatted document such as the ReadMe.md file.

The script [generate_version.sh](https://github.com/rexor12/holobot/blob/main/generate_version.sh) already describes this process on a high level, but it would be nice to have some graphical examples showing how to do it correctly with real branches.

To give a quick overview, it works as follows.

The head of the code is always the _main_ branch. This branch **must** always contain every commit from other branches as it represents the future, next generation release. Therefore, it **must** have every feature and bugfix. It can be considered as the playground, so major developments may take place here (such as refactoring the entire code-base) which may even introduce compatibility breaking changes. One should always do their best not to do so, but in case it is inevitable, this branch is the best place for that.

Other than that, semantic versioning is followed as described by [semver.org](https://semver.org), so the branches other than the _main_ branch follow conventions revolving around that.

Since work had been done already before moving to GitHub, the project started with version 1.0 which is represented by the branch _develop/1.0_. Here, the first number is the major number and the second number is the minor number. Major version 1 is the latest version as of writing this (apart from the _main_ branch that represents version 2.0).

The patch number is also represented in branch names starting from the first patch:
* main
* develop/1.2.**1**
* develop/1.2
* develop/1.1.**2**
* develop/1.1.**1**
* develop/1.1
* ... other branches ...
When creating a patch, you **must** create a new branch following the convention (e.g. _develop/1.2_ -> _develop/1.2.1_). To support development done by multiple people for a single patch, a branch should be created from the appropriate _develop_ branch, such as:
* main
* bugfix/develop/1.2.1/IncorrectArgumentValidation
* bugfix/develop/1.2.1/IncorrectExceptionHandling
* develop/1.2.1
* develop/1.2
* ... other branches ...
Once the development is done, these _bugfix_ branches are merged back to the parent branch _(develop/1.2.1_ in this case). Then, a release can be created - that is how people will know that the branch is now stable.

When a new feature needs to be implemented, the minor version has to increase, according to the semantic versioning specification. This is represented in this repository by creating a new branch from **the appropriate latest patch** branch with an increased minor number. Let's assume we have the following branches:
* main
* develop/1.2.1
* develop/1.2
* develop/1.1
* develop/1.0
In this case a new branch called _develop/1.3_ **has to** be created from the branch _develop/1.2.1_ as that one is the latest version for v1. This will ensure that all the other features and bugfixes are present in the new version, too.

To support development done by multiple people for a single feature, a branch should be created from the appropriate _develop_ branch. Let's assume we want to implement a feature sliced up to two parts - setting reminders and removing reminders - and we want the overall feature "Reminders" to be delivered in v1.3; this should be our branching:
* main
* feature/develop/1.3/RemoveReminderCommand
* feature/develop/1.3/SetReminderCommand
* develop/1.3
* develop/1.2.1
* ... other branches ...
Once development is done, these _feature_ branches are merged back to the parent branch (_develop/1.3_ in this case). Then, a release can be created - that is how people will know that the branch is now stable.

**Comment by rexor12** [comment]
Alright, that ended up not being a "quick overview". Instead of ReadMe.md, it might be worth considering to create a Wiki page (with graphical examples) to avoid bloating the "welcome" file.

**Comment by rexor12** [comment]
An update to this is that instead of branching from the highest develop branch, we should always branch from the _main_ branch. develop branches are the releases and new features shouldn't be developed on top of those; they should be developed on the head.

Branching from a develop branch should happen in one of two cases:
- when a bug needs to be fixed for that version (eg. develop/1.2 -> develop/1.2.1), or
- when a feature needs to be retrofited for some reason (eg. develop/1.2 -> develop/1.3)
